### PR TITLE
Use existing framework for backward dataflow analyses

### PR DESCRIPTION
**Pull request by ecstatic-morse**
This PR adds support for backward analyses to the dataflow framework and adds a new live variable analysis (based on the existing one in `librustc_mir/util/liveness.rs`). By adding these to the framework instead of having a separate API, all newly implemented backward dataflow analyses get cursors/visitors, `rustc_peek` tests, and graphviz visualizations for free. In the near-term, this makes it much easier to implement global dead-store elimination, and I believe that this will enable even more MIR optimizations in the future.

This PR makes many changes to the dataflow API, since some concepts and terminology only make sense in forward dataflow. Below is a list of the important changes.
- ~~`entry_set` -> `fixpoint` (the fixpoint for backward dataflow problems is after the block's terminator)~~
- `seek_{before,after}` -> `seek_{before,after}_primary_effect` (the unprefixed dataflow effect is now referred to as the "primary" effect instead of the "after" effect. The "before" effect remains the same, although I considered changing it to the "antecedent" effect. In both backward and forward dataflow, the "before" effect is applied prior to the "primary" effect. I feel very strongly that this is the correct choice, as it means consumers don't have to switch between `seek_before` and `seek_after` based on the direction of their analysis.
- `seek_after_assume_call_returns` is now gone. Users can use `ResultsCursor::apply_custom_effect` to emulate it.
- `visit_{statement,terminator}_exit` -> `visit_{statement,terminator}_after_primary_effect`
- `visit_{statement,terminator}` -> `visit_{statement,terminator}_before_primary_effect`

Implementing this also required refactoring the dataflow cursor implementation so it could work in both directions. This is a large percentage of the diff, since the cursor code is rather complex. The fact that the cursor is exhaustively tested in both directions should reassure whomever is unlucky enough to review this :rofl:.

In order to avoid computing the reverse CFG for forward dataflow analyses, I've added some hacks to the existing `mir::BodyAndCache` interface. I've requested changes to this interface that would let me implement this more efficiently.

r? @eddyb (feel free to reassign)
cc @rust-lang/wg-mir-opt

### CODE REVIEW

**Review by eddyb — APPROVED** [review]
LGTM at a glance

File: src/librustc_codegen_ssa/mir/analyze.rs
Comment (Centril) [inline src/librustc_codegen_ssa/mir/analyze.rs:272]:
```suggestion
            PlaceContext::MutatingUse(MutatingUseContext::Call | MutatingUseContext::Yield) => {
```

File: src/librustc_mir/dataflow/framework/backward.rs
Comment (Centril) [inline src/librustc_mir/dataflow/framework/backward.rs:41]: Hmm; maybe state this in the function docs?

File: src/librustc_mir/dataflow/framework/backward.rs
Comment (Centril) [inline src/librustc_mir/dataflow/framework/backward.rs:13]:
```suggestion
    range: RangeInclusive<EffectIndex>,
```
To take advantage of range notation?

File: src/librustc_mir/dataflow/impls/liveness.rs
Comment (Centril) [inline src/librustc_mir/dataflow/impls/liveness.rs:131]: Could use nested or-patterns here also.

**Comment by ecstatic-morse** [comment]
I'm now using an associated type instead of an associated const to indicate the direction of a dataflow analysis. This allowed me to remove some `if forward {} else {}` boilerplate. It also made the `visitor` module much cleaner.

@pnkfelix Let me know what I can do to ease review of this. An intra-block liveness analysis will make some MIR optimizations easier to implement, and I'd like to start experimenting with those. It will also be useful for #51003.

File: src/librustc_mir/dataflow/framework/tests.rs
Comment (ecstatic-morse) [inline src/librustc_mir/dataflow/framework/tests.rs:302]:
```suggestion
```

File: src/librustc_middle/mir/cache.rs
Comment (pnkfelix) [inline src/librustc_middle/mir/cache.rs:208]: This method doesn't have any documentation, so I cannot tell if it is meant to be an efficiency hack (which seems very unlikely, since an assert of `is_some()` should be incredibly cheap to check, no), or if it is something where at the call site (or at future expected call sites) you cannot guarantee that `cache.predecessors.is_some()`.

(I briefly tried to infer the intent by looking at the diff, so that I might suggest what comment to write here, or whether to advise removing the method entirely... From what I can tell, at the single call site where this is called, you are invoking `cache.ensure_predecessors` before this is invoked for the backwards flow case, so that will ensure that `cache.predecessors.is_some()`. But I'm guessing that predecessors is none for the forwards flow case...)

This is all a long winded way of saying that you should spell out in a doc-comment under what circumstances you expect `ReadOnlyBodyAndCache::new_unchecked` to be invoked rather than `ReadOnlyBodyAndCache::new`

Comment (ecstatic-morse) [reply src/librustc_middle/mir/cache.rs:208]: This is the part that will be removed after #71044 lands. Sorry. That should have been in the commit message for the `HACK` commit.

This is here because we want `join_state_into_successors` to have the same signature for both forward and backward analyses. Only backward analyses need to access the predecessor graph, but the current API demands that you must compute and cache the predecessor graph before creating a `ReadOnlyBodyAndCache` even if you don't actually use it.

FWIW, figuring this out how this was supposed to work caused me a great deal of consternation and was the motivation for #71044, which goes back to using `Body` everywhere.

Comment (pnkfelix) [reply src/librustc_middle/mir/cache.rs:208]: Maybe we should fold #71044 into this PR? I'll finish looking at this PR first before I look at that one.

**Comment by pnkfelix** [comment]
A general comment on one particular change:

>    * `entry_set` -> `fixpoint` (the fixpoint for backward dataflow problems is after the block's terminator)

I believe I understand your motivation for trying to find an alternative to "entry set", but I think this rename makes things more confusing.

I'm going to spell out my thinking explicitly here; little (or perhaps none) of this will be news  to you, @ecstatic-morse , but I want to write it down to justify why I think the above rename is ill chosen. You can skip to the Bikeshed section at the end if you like.

## Digression

### FIxed Points

A [fixed point](https://en.wikipedia.org/wiki/Fixed_point_(mathematics)) of a function *f*: D -> D is some input d that maps to itself (i.e. *f*(*d*) = *d*). In the context of a dataflow analysis, the domain D are the bitvectors that correspond to the state of the analysis at various points in the control flow of the program, and the analysis searches for a set of bitvectors such that the transfer functions attached to each block (and the operations for combining them at control flow merge points) end up causing no change to those bitvectors (i.e., each such bitvector is mapped to itself).

In short, to me (in the context of a dataflow analysis) a fixpoint is a point in the space of the state-space for the bitvectors themselves.

### Entry sets

In the data flow code itself, the *entry set* for a block is the bitvector associated with a particular point in the control flow of the block.

The term "fixpoint" for this thing seems inappropriate. The entry set may or may not be a bitvector that has reached a fixed point; in general that will depend on whether the whole analysis has reached a fixed point. And the entry set is definitely *used* in the code before the analysis has reached a fixed point.

In any case, it just seems like a mismatch to use a term that describes a point in the analysis state space to describe a part of the data structure that is most strongly associated with control flow.

The term "entry set" is meant to convey that this particular bitvector is the one that represents the analysis state on entry to the block.

As I said above, I can understand wanting to rename this, since one is likely to think that the entry point corresponds to when control flows *into* the block; but for a backwards analysis, these bitvectors stored for each block will be associated with the point where control flows *out* of the block (i.e., the point immediately after the block’s Terminator). If “entry” is too strongly correlated with (forward) control-flow entry, then “entry set” is a problem.

## Bikeshed

So the options here as I see them are either:

 * Leave the name `entry_set` as is, but try to clarify in the comments and/or surrounding docs that the word “entry” here means with respect to the direction of the analysis, *not* the direction of the actual control flow.

 * Keep some reference to the relationship with (control) flow, but try to generalize it beyond “entry” (e.g. `edge_state`, or `state_at_edge`

 * Abandon mentioning control flow in the name. (E. g. just `state` or `analysis_state`).

File: src/librustc_middle/mir/cache.rs
Comment (pnkfelix) [reply src/librustc_middle/mir/cache.rs:208]: Oh but perhaps you are concerned about performance issues with PR #71044, while you are confident that this PR should not inject performance problems, and thus it would be better to land them separately; is that a correct inference?

**Comment by ecstatic-morse** [comment]
@pnkfelix's concern is a valid one. As they hypothesize above, I found it hard to remember whether the "entry" of a block was in control-flow order (always before the first statement) or dataflow order (varies based on the direction of the analysis) while implementing this. I will point out that, once `iterate_to_fixpoint` is called, `fixpoint` *does* hold the dataflow state for each block *at* fixpoint. However, I see how using an abstract term to refer to a concrete field that is frequently mutating would cause confusion. I'm happy to change it.

I think that being very clear about the definition of "entry" within the dataflow framework is a valid approach. To me, this implies that we would refer to positions in control-flow order using "start" and "end" rather than "entry" and "exit". Part of the reason I chose "entry" and "exit" for control-flow order was a trivial one. Namely, the term "block start" would frequently appear next to the `BasicBlock` index `mir::START_BLOCK`.

So re: the bikeshed, I would choose to use "block start" and "block end" while preserving the name "entry set" for the dataflow state. The other alternatives are fine as well.

File: src/librustc_middle/mir/cache.rs
Comment (ecstatic-morse) [reply src/librustc_middle/mir/cache.rs:208]: To be clear, I was also concerned about regressions from this PR. I wanted to benchmark #71044 separately from this one, since the concerns/potential performance issues are very different in that PR (additional locking and slow serialization) than this one (renaming APIs and additional layers of abstraction).

Comment (ecstatic-morse) [reply src/librustc_middle/mir/cache.rs:208]: I've rebased on the latest master, which includes #71044. The HACK commit has been removed. Now both forward and backward dataflow analyses can use `Body` everywhere, and the reverse CFG will only be computed when necessary (see the most recent commit).

**Comment by pnkfelix** [comment]
> So re: the bikeshed, I would choose to use "block start" and "block end" while preserving the name "entry set" for the dataflow state. The other alternatives are also acceptable.

That sounds fine with me. My main concern is my inability to read expressions like `CursorPosition::fixpoint(block)`, because in my mind, there is no cursor position corresponding to a fixed point; it is, as I said above, a domain mismatch.

I don't know whether you plan to edit the commit series to include this change, or to add a commit that renames `fixpoint` where appropriate. (If you have time, I would prefer you edit the commit series, but I can understand if that is a serious burden.)

In either case, I'll continue on with my review to try to identify if there are any other problem areas so that you can incorporate that feedback as well.

**Comment by ecstatic-morse** [comment]
I'll edit the commit series (to evangelize a bit, I find [`git-absorb`](https://github.com/tummychow/git-absorb) is very helpful for this kind of rebase). Can I do this now? Or do you want me to avoid force pushing for a bit.

**Comment by pnkfelix** [comment]
> My main concern is my inability to read expressions like `CursorPosition::fixpoint(block)`, because in my mind, there is no cursor position corresponding to a fixed point; it is, as I said above, a domain mismatch.

 * Update: I will admit that, upon reading the comment "Resets the cursor to hold the fixpoint for the given basic block", that I can at least make some sense of what you are going for here.  

In short, it is treating the cursor as traversing the space of what the dataflow state is, versus having the cursor traverse the space of the control flow itself, right?

(I still think it will be *easier* to read if you go back to "entry"/"start"/etc or something similarly oriented around points in the control-flow, but I just wanted to point out that you were not on a completely wrong trajectory before.)

**Comment by pnkfelix** [comment]
> I'll edit the commit series (to evangelize, I find [`git-absorb`](https://github.com/tummychow/git-absorb) very helpful for this kind of rebase). Can I do this now? Or do you want me to avoid force pushing for a bit? I think force pushing might interfere with GitHub's "viewed files" feature for reviewing.

You can do it now. I think the benefit in making my review easier will make it pay off, at least for me.

(And in the meantime, I'll go find out what this `git-absorb` thing is...) :)

**Comment by ecstatic-morse** [comment]
Okay, that should be most of it. I also fixed the merge conflicts. I still need to add a section to the docs explaining the meaning of "block entry", but it should be reviewable at least.

File: src/librustc_mir/dataflow/framework/cursor.rs
Comment (pnkfelix) [inline src/librustc_mir/dataflow/framework/cursor.rs]: @ecstatic-morse did you overlook the case here?

Or are you implicitly arguing that the term "fixpoint" is appropriate here? (A potential viewpoint which I have attempted to understand in a comment elsewhere, but it isn't a viewpoint that I have been convinced to adopt...)

Comment (ecstatic-morse) [reply src/librustc_mir/dataflow/framework/cursor.rs]: I missed this case. Ugh.

Comment (ecstatic-morse) [reply src/librustc_mir/dataflow/framework/cursor.rs]: Ah this is just in a commit. Not the final PR. Presumably missed by git absorb

Comment (pnkfelix) [reply src/librustc_mir/dataflow/framework/cursor.rs]: (it seems like it must have been deleted by a later commit, now that I've compared against a full diff...)

anyway, thanks for confirming!

File: src/librustc_mir/dataflow/framework/cursor.rs
Comment (pnkfelix) [inline src/librustc_mir/dataflow/framework/cursor.rs:82]: likewise, here, you probably want to make sure this comment gets revised.

(especially since this sentence *does* have meaning, under my interpretation of "fixpoint", since the phrase "at fixpoint" could denote "the dataflow state at the (least) fixed point"; or, it could, as I suspect, carry your denotation of the so-called "fixpoint" control-flow position within the basic block (which is what I believe is the intent given the method name...)

Comment (ecstatic-morse) [reply src/librustc_mir/dataflow/framework/cursor.rs:82]: I seem to have switched the documentation for entry and start?

File: src/librustc_mir/dataflow/framework/engine.rs
Comment (pnkfelix) [inline src/librustc_mir/dataflow/framework/engine.rs:210]: This is an interesting question.

I tried to do some source code archeology to determine when/why we add all blocks, including unreachable ones, to the initial dirty queue.

As far as I can tell, that has been present since the dirty list was introduced in PR #51900. (Before that, we did something even more naive and traversed the entire set of basic blocks on every iteration.)

We have had cases in the past where it was important to include all MIR basic blocks because of source code lints that relied on such analysis data being present even for unreachable code. But I don't know if any of our lints are using the results of dataflow...?

File: src/test/ui/mir-dataflow/liveness-ptr.rs
Comment (pnkfelix) [inline src/test/ui/mir-dataflow/liveness-ptr.rs:18]: 🎉

**Comment by pnkfelix** [comment]
Okay I've finished looking at the commit series.

Looks great! Thanks for all the work you've put into this, @ecstatic-morse 

r=me once the you've addressed the overlooked case of `CursorPosition::fixpoint` that we discussed above.

File: src/librustc_mir/dataflow/framework/engine.rs
Comment (ecstatic-morse) [reply src/librustc_mir/dataflow/framework/engine.rs:210]: In an initial version of #64566, I added an `assert` that fired when dataflow was run on a CFG with unreachable basic blocks. Most of the time (MIR borrowck, drop elaboration), we do dataflow after a `SimplifyCfg` pass so this is a non-issue, but the generator transform pass was triggering the assert, so I added that last loop in to preserve existing behavior.

It would be nice to `panic` when we tried to *inspect* the dataflow results in an unreachable basic block, since it likely won't be what the user is expecting even if we continue adding them to the dirty queue initially (e.g. it won't have the `START_BLOCK` state propagated into it).

File: src/librustc_mir/dataflow/framework/cursor.rs
Comment (ecstatic-morse) [reply src/librustc_mir/dataflow/framework/cursor.rs]: Okay, I've changed this family of methods to use the phrase "dataflow state at fixpoint" when seeking to the dataflow entry point. I've also tried to make it clear that `seek_to_block_{start,end}` are referring to the control flow position, not the dataflow one by explicitly mentioning "first statement" and "terminator". These changes are isolated in the last commit.

**Comment by ecstatic-morse** [comment]
Okay, I rebased away the mention of `fixpoint` in the giant commit and made some docs changes in the last commit in the series to address one of your notes. I'll wait until CI is green to send to the queue.

**Pull request merged**
